### PR TITLE
schedule/placement/fit: remove the recursion in pick peers for on rule

### LINKED ISSUE #5268: PlacementRule: could we use binary number to remove the recursion in pick peers for rules
## Enhancement Task

[https://github.com/tikv/pd/blob/c8775b6176a03eb25cb1622fc46ba1d496872df7/server/schedule/placement/fit.go#L234-L254](https://github.com/tikv/pd/blob/c8775b6176a03eb25cb1622fc46ba1d496872df7/server/schedule/placement/fit.go#L234-L254)

### PR DESCRIPTION
**Pull request by AndreMouche**
Signed-off-by: shirly <AndreMouche@126.com>

### What problem does this PR solve?

Issue Number: Close #5268 

### Check List

Tests

- Unit test

- benchmark

#### current branch

```
➜  placement git:(improve_fit) go test -benchmem -run=^$ -bench ^BenchmarkFitRegionWithMoreRulesAndStoreLabels -benchtime=100x
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/schedule/placement
cpu: VirtualApple @ 2.50GHz
BenchmarkFitRegionWithMoreRulesAndStoreLabels-8              100          24242994 ns/op         1334621 B/op      55080 allocs/op
PASS
ok      github.com/tikv/pd/server/schedule/placement    3.437s
```

#### Master

```
➜  placement git:(master) go test -benchmem -run=^$ -bench ^BenchmarkFitRegionWithMoreRulesAndStoreLabels -benchtime=100x
goos: darwin
goarch: amd64
pkg: github.com/tikv/pd/server/schedule/placement
cpu: VirtualApple @ 2.50GHz
BenchmarkFitRegionWithMoreRulesAndStoreLabels-8              100          48844034 ns/op         2627012 B/op     107788 allocs/op
PASS
ok      github.com/tikv/pd/server/schedule/placement    6.110s
```

### Release note

```release-note
 None.
```

### CODE REVIEW

File: server/schedule/placement/fit.go [hunk 1]
```diff
@@ -227,34 +227,73 @@ func (w *fitWorker) fitRule(index int) bool {
...
-	// make sure the left number of candidates should be enough.
-	indexLimit := len(candidates) - (count - len(selected))
-	for i := 0; i <= indexLimit; i++ {
-		p := candidates[i]
-		p.selected = true
-		better = w.enumPeers(candidates[i+1:], append(selected, p), index, count) || better
-		p.selected = false
+
+	for binaryInt := (1<<count - 1); binaryInt <= limit; binaryInt++ {
+		// there should be exactly `count` number in current binary number `m`
```
Comment (rleungx) [inline server/schedule/placement/fit.go:244]: What does the `m` here mean?

**Comment by AndreMouche** [comment]
> Mostly LGTM, but I suggest adding some tests for `seletedBitEqualsTo` and `pickPeersFromBinaryInt`

I've used `bits.OnesCount(binaryInt)` instead `seletedBitEqualsTo` since it is faster. Meanwhile, the test  for ``pickPeersFromBinaryInt`` was added

File: server/schedule/placement/fit.go [hunk 2]
```diff
@@ -227,34 +228,60 @@ func (w *fitWorker) fitRule(index int) bool {
...
-		p.selected = false
+	limit := uint(1<<len(candidates) - 1)
+	binaryInt := uint(1<<count - 1)
+	for ; binaryInt <= limit; binaryInt++ {
+		// there should be exactly `count` number in current binary number `binaryInt`
+		if bits.OnesCount(binaryInt) != count {
+			continue
+		}
+		selected := pickPeersFromBinaryInt(candidates, binaryInt)
+		better = w.compareBest(selected, index) || better
```
Comment (rleungx) [inline server/schedule/placement/fit.go:251]: Do we need `|| better`?

Comment (AndreMouche) [reply server/schedule/placement/fit.go:251]: yes, better means we find a better solution for current rule[index] with candidates. 
it is useful that when we find two solutions for rule[index] are equal, we need to compare the rule[index+1] with the corresponding candidates.

**Review by rleungx — APPROVED** [review]

**Comment by nolouch** [comment]
PTAL @CabinfeverB

File: server/schedule/placement/fit_test.go [hunk 3]
```diff
@@ -187,3 +187,36 @@ func TestIsolationScore(t *testing.T) {
...
+	for _, c := range testCases {
+		binaryNumber, err := strconv.ParseUint(c.binary, 2, 64)
+		re.Nil(err)
+		selected := pickPeersFromBinaryInt(candidates, uint(binaryNumber))
+		re.Len(selected, len(c.expectedPeers))
+		for id := 0; id < len(selected); id++ {
+			re.Equal(selected[id].Id, c.expectedPeers[id])
+		}
+	}
+
```
Comment (CabinfeverB) [inline server/schedule/placement/fit_test.go:221]: remove this line

**Review by CabinfeverB — APPROVED** [review]

File: server/schedule/placement/fit_test.go [hunk 4]
```diff
@@ -187,3 +187,35 @@ func TestIsolationScore(t *testing.T) {
...
+		{"101", []uint64{1, 3}},
+		{"111", []uint64{1, 2, 3}},
+		{"1011", []uint64{1, 2, 4}},
+		{"100011", []uint64{1, 2, 6}},
+		{"1000001111", []uint64{1, 2, 3, 4, 10}},
+	}
+
+	for _, c := range testCases {
+		binaryNumber, err := strconv.ParseUint(c.binary, 2, 64)
+		re.Nil(err)
```
Comment (JmPotato) [inline server/schedule/placement/fit_test.go:214]:
```suggestion
		re.NoError(err)
```

**Pull request merged**